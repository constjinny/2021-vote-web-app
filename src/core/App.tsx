import { ReactElement, useEffect } from "react";
import styled from "@emotion/styled";
import { CommAPI } from "../api/sample.api";

import Routes from "../routes/Routes";

function App(): ReactElement {
  useEffect(() => {
    // NOTI: 임시 API사용으로 최초 진입 시 세션에 데이터 별도 세팅
    CommAPI.setInitData();
  }, []);

  return (
    <WrapStyle>
      <Routes />
    </WrapStyle>
  );
}

export default App;

const WrapStyle = styled.div`
  height: 100%;
`;
