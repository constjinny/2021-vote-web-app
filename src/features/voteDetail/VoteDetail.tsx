import { ReactElement } from "react";
import styled from "@emotion/styled";

export function VoteDetail(): ReactElement {
  return <VoteDetailStyle>VoteDetail</VoteDetailStyle>;
}

const VoteDetailStyle = styled.div``;
