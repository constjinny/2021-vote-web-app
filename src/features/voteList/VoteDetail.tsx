import { ReactElement, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import styled from "@emotion/styled";
import { isEmpty } from "lodash";

import { voteListAction, voteListSelector } from "./VoteListSlice";
import { IVoteItemProps, IVoteOptionProps } from "./type";
import { Button, ButtonTheme } from "../../components/button";

export function VoteDetail({ data }: { data: IVoteItemProps }): ReactElement {
  const dispatch = useDispatch();
  const userId = useSelector(voteListSelector.selectUserId);
  const [selectValueId, setSelectValueId] = useState("");
  const { isOpen, isClose, openDate, voteOptions, isSelectId, voteId } = data;

  const handleResultOption = () =>
    dispatch(
      voteListAction.updateVoteOptionCount({
        userId,
        voteId,
        optionId: selectValueId,
      })
    );

  const Description = isSelectId ? (
    <p>🗳 이미 투표를 하셨습니다.</p>
  ) : !isOpen ? (
    <p>🗳 {openDate} 부터 투표 가능합니다.</p>
  ) : null;

  return (
    <VoteDetailStyle>
      <OptionListStyle>
        {Description && Description}

        {voteOptions.map((option: IVoteOptionProps, index: number) => {
          const { optionId, optionValue, count } = option;
          const isSelect =
            optionId === selectValueId || optionId === isSelectId;
          const isDisabled = !isOpen || isClose || !isEmpty(isSelectId);

          const handleSelectValue = () =>
            !isDisabled && setSelectValueId(optionId);

          return (
            <OptionStyle
              key={optionId}
              isSelect={isSelect}
              isDisabled={isDisabled}
              onClick={handleSelectValue}
            >
              <span>{index + 1}</span>

              <div>
                {optionValue}
                <span>{`${count} 명`}</span>
              </div>
            </OptionStyle>
          );
        })}
      </OptionListStyle>

      {isOpen && !isClose && isEmpty(isSelectId) && (
        <ButtonGroupStyle>
          <Button
            theme={ButtonTheme.Full}
            onClick={handleResultOption}
            isDisabled={isEmpty(selectValueId)}
          >
            투표하기
          </Button>
        </ButtonGroupStyle>
      )}
    </VoteDetailStyle>
  );
}

const VoteDetailStyle = styled.div`
  margin-top: 20px;
  padding: 20px;
  border-radius: 5px;
  border: 1px solid #cacaca;
  background: #fff;
  box-sizing: border-box;
`;

const OptionListStyle = styled.ul`
  > p {
    margin-bottom: 10px;
  }
`;

const OptionStyle = styled.li<{ isSelect: boolean; isDisabled: boolean }>`
  overflow: hidden;
  position: relative;
  margin-top: 10px;
  padding: 10px 20px 10px 50px;
  font-size: 14px;
  border-radius: 5px;
  border: 1px solid #cacaca;
  ${({ isSelect }) => isSelect && "border-color: #5ea9e7"};
  ${({ isDisabled }) => isDisabled && "background-color: #eeeeee;"};
  &:hover {
    ${({ isDisabled }) =>
      !isDisabled &&
      "cursor: pointer; border-color: #457aa7; box-shadow: rgba(149, 157, 165, 0.9) 0px 10px 24px;"};
  }
  > span {
    position: absolute;
    top: 10px;
    left: 20px;
    width: 20px;
    height: 20px;
    text-align: center;
    font-weight: bold;
    line-height: 20px;
    border-radius: 50%;
    background-color: ${({ isSelect }) => (isSelect ? "#5ea9e7" : "#cacaca")};
    color: #fff;
  }
  > div {
    display: flex;
    justify-content: space-between;
    align-items: center;
    > span {
      display: inline-block;
    }
  }
`;

const ButtonGroupStyle = styled.div`
  display: flex;
  justify-content: space-between;
  margin-top: 20px;
  padding-top: 20px;
  border-top: 1px solid #cacaca;
`;
