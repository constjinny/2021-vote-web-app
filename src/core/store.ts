import { configureStore, Action } from "@reduxjs/toolkit";
import { combineReducers } from "@reduxjs/toolkit";
import { ThunkAction } from "redux-thunk";
import { useDispatch } from "react-redux";

import { voteListReducer } from "../features/voteList";
import { voteWriteReducer } from "../features/voteCreate";

export const rootReducer = combineReducers({
  voteListReducer,
  voteWriteReducer,
});

export type IRootState = ReturnType<typeof rootReducer>;

const store = configureStore({
  reducer: rootReducer,
});

export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();
export type AppThunk = ThunkAction<void, IRootState, null, Action<string>>;

export default store;
