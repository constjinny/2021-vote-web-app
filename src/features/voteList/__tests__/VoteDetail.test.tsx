import { configureStore } from "@reduxjs/toolkit";
import { rootReducer } from "../../../core/store";
import { Provider } from "react-redux";
import { shallow } from "enzyme";

import { VoteDetail } from "../VoteDetail";

const testData = {
  author: {
    id: "jinny",
    nickName: "jinny",
    profileUrl: "jinny_profile.png",
  },
  closeDate: "2021년 6월 19일",
  isClose: false,
  isEdit: true,
  isOpen: false,
  isResultId: "",
  isSelectId: "",
  openDate: "2021년 6월 18일",
  voteId: "b830191fdi",
  voteName: "테스트 투표",
  voteOptions: [
    {
      count: 0,
      optionId: "b830191fdi-item1",
      optionValue: "투표 항목",
      selectedIds: [],
    },
  ],
};

describe("VoteDetail", () => {
  const store = configureStore({
    reducer: rootReducer,
  });
  let component: any = null;

  it("renders VoteDetail component", () => {
    component = shallow(
      <Provider store={store}>
        <VoteDetail data={testData} />
      </Provider>
    );
  });

  it("matches snapshot", () => {
    expect(component).toMatchSnapshot();
  });
});
