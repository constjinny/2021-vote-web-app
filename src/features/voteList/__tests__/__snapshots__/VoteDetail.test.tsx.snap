// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoteDetail matches snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <VoteDetail
    data={
      Object {
        "author": Object {
          "id": "jinny",
          "nickName": "jinny",
          "profileUrl": "jinny_profile.png",
        },
        "closeDate": "2021년 6월 19일",
        "isClose": false,
        "isEdit": true,
        "isOpen": false,
        "isResultId": "",
        "isSelectId": "",
        "openDate": "2021년 6월 18일",
        "voteId": "b830191fdi",
        "voteName": "테스트 투표",
        "voteOptions": Array [
          Object {
            "count": 0,
            "optionId": "b830191fdi-item1",
            "optionValue": "투표 항목",
            "selectedIds": Array [],
          },
        ],
      }
    }
  />
</ContextProvider>
`;
